consensus/event_loop.go

```diff
@@ -0,0 +1,854 @@
+	}
+
+	// Schedule the next crank timer task. This will run with currentView param.
+	fc.crankTimerTask.Schedule(fc.crankTimerInterval, fc.currentView, fc.onCrankTimerTaskExecuted)
+
+	// Schedule the next timeout task. This will run with currentView param.
+	fc.nextTimeoutTask.Schedule(timeoutDuration, fc.currentView, fc.onTimeoutScheduledTaskExecuted)
+}
+
+// When this function is triggered, it means that we have reached the crank timer
+// time ETA for blockConstructionView. If we have a QC or timeout QC for the view, then we
+// signal the server.
+func (fc *FastHotStuffEventLoop) onCrankTimerTaskExecuted(blockConstructionView uint64) {
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Check if the event loop is running. If it's not running, then there's nothing
+	// to do here.
+	if fc.status != eventLoopStatusRunning {
+		return
+	}
+
+	// Check for race conditions where the view advanced at the exact moment this task began
+	// or we have already signaled for this view. If so, then there's nothing to do here.
+	if fc.currentView != blockConstructionView {
+		return
+	}
+
+	// Mark that the crank timer has elapsed
+	fc.hasCrankTimerRunForCurrentView = true
+
+	// Check if the conditions are met to construct a QC from votes for the chain tip. If so,
+	// we send a signal to the server and cancel the crank timer task. The server will
+	// reschedule the task when it advances the view.
+	if voteQCEvent := fc.tryConstructVoteQCInCurrentView(); voteQCEvent != nil {
+		// Signal the server that we can construct a QC for the chain tip
+		fc.hasConstructedQCInCurrentView = true
+		fc.Events <- voteQCEvent
+		return
+	}
+
+	// Check if we have enough timeouts to build an aggregate QC for the previous view. If so,
+	// we send a signal to the server and cancel all scheduled tasks.
+	if timeoutQCEvent := fc.tryConstructTimeoutQCInCurrentView(); timeoutQCEvent != nil {
+		// Signal the server that we can construct a timeout QC for the current view
+		fc.hasConstructedQCInCurrentView = true
+		fc.Events <- timeoutQCEvent
+		return
+	}
+}
+
+// tryConstructVoteQCInCurrentView is a helper function that attempts to construct a QC for the tip block
+// so that it can be proposed in a block in the current view. The function internally performs all view and vote
+// validations to ensure that the resulting QC is valid. If a QC can be constructed, the function returns
+// the signers list and aggregate signature that can be used to construct the QC.
+//
+// This function must be called while holding the event loop's lock.
+func (fc *FastHotStuffEventLoop) tryConstructVoteQCInCurrentView() *FastHotStuffEvent {
+	// If currentView != tipBlock.View + 1, then we have timed out at some point, and can no longer
+	// construct a block with a QC of votes for the tip block.
+	tipBlock := fc.tip.block
+	if fc.currentView != tipBlock.GetView()+1 {
+		return nil
+	}
+
+	// Fetch the validator set at the tip.
+	validatorSet := fc.tip.validatorSet
```

Comment:
Nit: Note this variable name if you decide to change validatorSet -> validatorList